onnxruntime/core/providers/cpu/math/top_k.cc

```diff
@@ -45,92 +47,175 @@ static int64_t SizeFromDim(size_t k, const vector<int64_t>& dims) {
 }
 
 template <typename T>
-struct ValueCmp {
+struct GreaterValueCmp {
   bool operator()(
       const pair<T, int64_t>& lhs,
       const pair<T, int64_t>& rhs) {
     return (
         lhs.first > rhs.first ||
+        // when values are equal, we want lhs to get higher "priority"
+        // if its corresponding index comes first (i.e.) is lower
         (lhs.first == rhs.first && lhs.second < rhs.second));
   }
 };
 
-// Core TopK implementation
-Status TopKImpl(OpKernelContext* p_op_kernel_context, const Tensor* X, const int axis, const unsigned k) {
-
-  const vector<int64_t>& in_dims = X->Shape().GetDims();
-  // Will return axis_ as is if positive or fixes it in case it is negative
-  auto axis_parsed = HandleNegativeAxis(axis, in_dims.size());
-  // Check to ensure k is within the bounds of what is available in that specific axis
-  if (in_dims.at(axis_parsed) < k) {
-    ostringstream err_msg;
-    err_msg << "k argment [" << k << "] should not be greater than specified axis dim value [" << in_dims.at(axis_parsed) << "]";
-    return Status(common::ONNXRUNTIME, common::FAIL, err_msg.str());
+template <typename T>
+struct LesserValueCmp {
+  bool operator()(
+      const pair<T, int64_t>& lhs,
+      const pair<T, int64_t>& rhs) {
+    return (
+        lhs.first < rhs.first ||
+        // when values are equal, we want lhs to get higher "priority"
+        // if its corresponding index comes first (i.e.) is lower
+        (lhs.first == rhs.first && lhs.second < rhs.second));
   }
+};
 
-  if (k == 0) {
-    vector<int64_t> out_dims = in_dims;
-    out_dims[axis_parsed] = 0;
-    p_op_kernel_context->Output(0, out_dims);
-    p_op_kernel_context->Output(1, out_dims);
-    return Status::OK();
-  }
+// Static helpers that implement the core logic for each of the 'TopK' operator flavor
 
-  const int64_t rows = SizeToDim(axis_parsed, in_dims);
-  const int64_t cols = X->Shape().Size() / rows;
+// Given an input tensor 'input' and metadata values - 'k' and 'axis_parsed',
+// this method will extract the sorted top k largest/smallest elements and place them in the output tensor 'values'
+// along with the metadata output 'indices'
+template <bool largest, bool sorted, class Comparator>
+static void extract_top_k_elements(const Tensor* input, const std::vector<int64_t>& input_shape,
```

Comment:
>const std::vector<int64_t>& input_shape [](start = 56, length = 39)

Why use the vector instead of TensorShape which already provides methods to get sizes to/from dimensions?

Reply:
Fixed